### ISSUE TITLE
`vetiver_write_docker()` missing sys_reqs 

### ISSUE THREAD

**Issue by michkam89** [rstudio/vetiver-r#150]
Hi!

I was exploring `vetiver_write_docker()` and I've found that sometimes it does not identify all system requirements and the build of the image fails. 

I think that this might be related to the fact that what is passed to `glue_sys_reqs()` [here](https://github.com/rstudio/vetiver-r/blob/99f82bdedc2e87adf4439ad0c604afb564d307fb/R/write-docker.R#L71) are only `unique(c(docker_pkgs, vetiver_model$metadata$required_pkgs))`  and some packages that are added to the `vetiver_renv.lock` are missed. 

Example can be given when I have `{gert}` package in `vetiver_renv.lock` and this is the part of the `renv::restore()` error message:

```
#9 362.6 Installing gert [1.6.0] ...
#9 363.9        FAILED
#9 363.9 Error installing package 'gert':
#9 363.9 ================================
#9 363.9
#9 363.9 * installing *source* package ‘gert’ ...
#9 363.9 ** package ‘gert’ successfully unpacked and MD5 sums checked
#9 363.9 ** using staged installation
#9 363.9 > curl::curl_download("https://r-lib.github.io/gert/get-libgit2-linux.sh","get-libgit2-linux.sh")
#9 363.9 Error in curl::curl_download("https://r-lib.github.io/gert/get-libgit2-linux.sh",  :
#9 363.9   SSL certificate problem: self signed certificate in certificate chain
#9 363.9 Execution halted
#9 363.9 Using PKG_CFLAGS=
#9 363.9 Using PKG_LIBS=-lgit2
#9 363.9 ----------------------------- ANTICONF -------------------------------
#9 363.9 Configuration failed to find libgit2 library. Try installing:
#9 363.9  * brew: libgit2 (MacOS)
#9 363.9  * deb: libgit2-dev (Debian, Ubuntu, etc)
#9 363.9  * rpm: libgit2-devel (Fedora, CentOS, RHEL)
#9 363.9 If libgit2 is already installed, check that 'pkg-config' is in your
#9 363.9 PATH and PKG_CONFIG_PATH contains a libgit2.pc file. If pkg-config
#9 363.9 is unavailable you can set INCLUDE_DIR and LIB_DIR manually via:
#9 363.9 R CMD INSTALL --configure-vars='INCLUDE_DIR=... LIB_DIR=...'
#9 363.9 -------------------------- [ERROR MESSAGE] ---------------------------
#9 363.9 <stdin>:1:10: fatal error: git2.h: No such file or directory
#9 363.9 compilation terminated.
#9 363.9 ----------------------------------------------------------------------
#9 363.9 ERROR: configuration failed for package ‘gert’
#9 363.9 * removing ‘/usr/local/lib/R/site-library/.renv/1/gert’
#9 363.9 Error: install of package 'gert' failed [error code 1]
#9 363.9 Traceback (most recent calls last):
#9 363.9 12: renv::restore()
#9 363.9 11: renv_restore_run_actions(project, diff, current, lockfile, rebuild)
#9 363.9 10: renv_install_impl(records)
#9 363.9  9: renv_install_staged(records)
#9 363.9  8: renv_install_default(records)
#9 363.9  7: handler(package, renv_install_package(record))
#9 363.9  6: renv_install_package(record)
#9 363.9  5: withCallingHandlers(renv_install_package_impl(record), error = function(e) {
#9 363.9         vwritef("\tFAILED")
#9 363.9         writef(e$output)
#9 363.9     })
#9 363.9  4: renv_install_package_impl(record)
#9 363.9  3: r_cmd_install(package, path)
#9 363.9  2: r_exec_error(package, output, "install", status)
#9 363.9  1: stop(error)
#9 364.0 Execution halted
------
executor failed running [/bin/sh -c Rscript -e "renv::restore()"]: exit code: 1
```

Was it intentional eg. to limit the amount of packages or system requirements? 

Thanks!

**Comment by juliasilge** [comment]
Can you give me a bit more info about how you are getting into this situation? For example, what model are you using that puts gert into the `vetiver_renv.lock` file? I haven't run into this with my own demos or testing.

The goal of this function is to get the set of packages needed here:

https://github.com/rstudio/vetiver-r/blob/99f82bdedc2e87adf4439ad0c604afb564d307fb/R/write-docker.R#L61-L62

And then use those packages both in [writing the lock file](https://github.com/rstudio/vetiver-r/blob/99f82bdedc2e87adf4439ad0c604afb564d307fb/R/write-docker.R#L66) and [getting the system requirements](https://github.com/rstudio/vetiver-r/blob/99f82bdedc2e87adf4439ad0c604afb564d307fb/R/write-docker.R#L71).

**Comment by michkam89** [comment]
Yes sure. I used one of the examples with `ranger` but needed to add `paws.storage` as I want to create `board_s3` according to #149 .

```r
# creating model
library(parsnip)
library(workflows)
data(Sacramento, package = "modeldata")

rf_spec <- rand_forest(mode = "regression")
rf_form <- price ~ type + sqft + beds + baths

rf_fit <- 
    workflow(rf_form, rf_spec) %>%
    fit(Sacramento)

library(vetiver)
v <- vetiver_model(rf_fit, "sacramento_rf")

# create board 
my_s3_board_vetiver <- pins::board_s3(
  bucket = "<some name>", 
  prefix = "test-vetiver/", 
  region = "eu-central-1"
  )

# add paws 
v$metadata$required_pkgs <- c(v$metadata$required_pkgs, "paws.storage")

# write model
my_s3_board_vetiver %>% 
  vetiver_pin_write(
    vetiver_model = v,
    ServerSideEncryption = "aws:kms", 
    SSEKMSKeyId = "<some key id>")

# plumber and docker
vetiver_write_plumber(my_s3_board_vetiver, name = "sacramento_rf", rsconnect = FALSE)
vetiver_write_docker(v)
```

Additionally this is the `renv` console output I get from `vetiver_write_docker()`:
```
The following package(s) will be updated in the lockfile:

# CRAN ===============================
- MASS           [* -> 7.3-54]
- Matrix         [* -> 1.3-4]
- R6             [* -> 2.5.1]
- RColorBrewer   [* -> 1.1-3]
- Rcpp           [* -> 1.0.8.3]
- RcppEigen      [* -> 0.3.3.9.2]
- askpass        [* -> 1.1]
- base64enc      [* -> 0.1-3]
- bit            [* -> 4.0.4]
- bit64          [* -> 4.0.5]
- bundle         [* -> 0.1.0]
- butcher        [* -> 0.1.5]
- cli            [* -> 3.3.0]
- clipr          [* -> 0.8.0]
- codetools      [* -> 0.2-18]
- colorspace     [* -> 2.0-3]
- cpp11          [* -> 0.4.2]
- crayon         [* -> 1.5.1]
- credentials    [* -> 1.3.2]
- curl           [* -> 4.3.2]
- desc           [* -> 1.4.1]
- digest         [* -> 0.6.29]
- dplyr          [* -> 1.0.9]
- ellipsis       [* -> 0.3.2]
- fansi          [* -> 1.0.3]
- farver         [* -> 2.1.0]
- filelock       [* -> 1.0.2]
- fs             [* -> 1.5.2]
- generics       [* -> 0.1.2]
- gert           [* -> 1.6.0]
- ggplot2        [* -> 3.3.6]
- gh             [* -> 1.3.0]
- gitcreds       [* -> 0.1.1]
- globals        [* -> 0.15.0]
- glue           [* -> 1.6.2]
- gtable         [* -> 0.3.0]
- hardhat        [* -> 1.0.0]
- hms            [* -> 1.1.1]
- httpuv         [* -> 1.6.5]
- httr           [* -> 1.4.3]
- ini            [* -> 0.3.1]
- isoband        [* -> 0.2.5]
- jsonlite       [* -> 1.8.0]
- labeling       [* -> 0.4.2]
- later          [* -> 1.3.0]
- lattice        [* -> 0.20-45]
- lifecycle      [* -> 1.0.1]
- lobstr         [* -> 1.1.1]
- magrittr       [* -> 2.0.3]
- mgcv           [* -> 1.8-38]
- mime           [* -> 0.12]
- munsell        [* -> 0.5.0]
- nlme           [* -> 3.1-153]
- openssl        [* -> 2.0.2]
- parsnip        [* -> 0.2.1]
- paws.common    [* -> 0.3.17]
- paws.storage   [* -> 0.1.12]
- pillar         [* -> 1.7.0]
- pins           [* -> 1.0.3]
- pkgconfig      [* -> 2.0.3]
- plumber        [* -> 1.1.0]
- prettyunits    [* -> 1.1.1]
- progress       [* -> 1.2.2]
- promises       [* -> 1.2.0.1]
- purrr          [* -> 0.3.4]
- ranger         [* -> 0.13.1]
- rapidoc        [* -> 8.4.3]
- rappdirs       [* -> 0.3.3]
- readr          [* -> 2.1.2]
- renv           [* -> 0.15.5]
- rlang          [* -> 1.0.2]
- rprojroot      [* -> 2.0.3]
- rstudioapi     [* -> 0.13]
- scales         [* -> 1.2.0]
- sodium         [* -> 1.2.0]
- stringi        [* -> 1.7.6]
- swagger        [* -> 3.33.1]
- sys            [* -> 3.4]
- tibble         [* -> 3.1.7]
- tidyr          [* -> 1.2.0]
- tidyselect     [* -> 1.1.2]
- tzdb           [* -> 0.3.0]
- usethis        [* -> 2.1.6]
- utf8           [* -> 1.2.2]
- vctrs          [* -> 0.4.1]
- vetiver        [* -> 0.1.8]
- viridisLite    [* -> 0.4.0]
- vroom          [* -> 1.5.7]
- webutils       [* -> 1.1]
- whisker        [* -> 0.4]
- withr          [* -> 2.5.0]
- workflows      [* -> 0.2.6]
- xml2           [* -> 1.3.3]
- yaml           [* -> 2.3.5]
- zip            [* -> 2.2.0]
```

while the `pkgs` vector passed to `glue_sys_reqs()` is as follows:
```r
pkgs
[1] "pins"         "plumber"      "rapidoc"      "vetiver"      "renv"        
[6] "parsnip"      "ranger"       "workflows"    "paws.storage"
```

The summary of the issue I see, is that there are more packages in `renv.lock` that may contain additional system requirements and fail during `renv::restore()` when building the image :)

In code that would manifest in this small change:
```r
    pkgs <- unique(c(docker_pkgs, vetiver_model$metadata$required_pkgs))
    pkgs <- setdiff(pkgs, drop_pkgs)
    lockfile_pkgs <- renv::snapshot(
        project = path,
        lockfile = lockfile,
        packages = pkgs,
        prompt = FALSE,
        force = TRUE
    )
    plumber_file <- fs::path_rel(plumber_file)

    pkgs <- unique(c(pkgs, names(lockfile_pkgs$Packages)))
    sys_reqs <- glue_sys_reqs(pkgs)
```

**Comment by michkam89** [comment]
In the meantime I executed all those steps on my other machine as well as in docker container and I received different `vetiver_renv.lock` file (eg. without `{gert}`). With this, I've built the image successfully.

That was probably caused by the fact that I had my own `renv.lock` file in this project and some packages were downloaded from quite old cran snapshot or something...

Nevertheless, I think the question about searching system requirements is still valid, as under some circumstances such a situation may happen :)
